pinot-controller/src/main/java/org/apache/pinot/controller/recommender/realtime/provisioning/MemoryEstimator.java

```diff
@@ -0,0 +1,589 @@
+      throws IOException {
+    // We don't want the stats history to get updated from all our dummy runs
+    // So we copy over the original stats history every time we start
+    File statsFileCopy = new File(_workingDir, STATS_FILE_COPY_NAME);
+    FileUtils.copyFile(statsFile, statsFileCopy);
+    RealtimeSegmentStatsHistory statsHistory;
+    try {
+      statsHistory = RealtimeSegmentStatsHistory.deserialzeFrom(statsFileCopy);
+    } catch (IOException | ClassNotFoundException e) {
+      throw new RuntimeException(
+          "Exception when deserializing stats history from stats file " + statsFileCopy.getAbsolutePath(), e);
+    }
+    RealtimeIndexOffHeapMemoryManager memoryManager = new DirectMemoryManager(_segmentMetadata.getName());
+    SegmentZKMetadata segmentZKMetadata = getSegmentZKMetadata(_segmentMetadata, totalDocs);
+
+    RealtimeSegmentConfig.Builder realtimeSegmentConfigBuilder =
+        new RealtimeSegmentConfig.Builder().setTableNameWithType(_tableNameWithType)
+            .setSegmentName(_segmentMetadata.getName()).setStreamName(_tableNameWithType)
+            .setSchema(_segmentMetadata.getSchema()).setCapacity(totalDocs).setAvgNumMultiValues(_avgMultiValues)
+            .setNoDictionaryColumns(_noDictionaryColumns).setVarLengthDictionaryColumns(_varLengthDictionaryColumns)
+            .setInvertedIndexColumns(_invertedIndexColumns).setSegmentZKMetadata(segmentZKMetadata).setOffHeap(true)
+            .setMemoryManager(memoryManager).setStatsHistory(statsHistory);
+
+    // create mutable segment impl
+    MutableSegmentImpl mutableSegmentImpl = new MutableSegmentImpl(realtimeSegmentConfigBuilder.build(), null);
+    long memoryForConsumingSegmentPerPartition = memoryManager.getTotalAllocatedBytes();
+    mutableSegmentImpl.destroy();
+    FileUtils.deleteQuietly(statsFileCopy);
+    return memoryForConsumingSegmentPerPartition;
+  }
+
+  /**
+   * Gets the average num multivalues across all multi value columns in the data
+   * @return
+   */
+  private int getAvgMultiValues() {
+    int avgMultiValues = 0;
+    Set<String> multiValueColumns =
+        _segmentMetadata.getSchema().getAllFieldSpecs().stream().filter(fieldSpec -> !fieldSpec.isSingleValueField())
+            .map(FieldSpec::getName).collect(Collectors.toSet());
+
+    if (!multiValueColumns.isEmpty()) {
+
+      int numValues = 0;
+      long multiValuesSum = 0;
+      try {
+        PinotSegmentRecordReader segmentRecordReader = new PinotSegmentRecordReader(_sampleCompletedSegment);
+        GenericRow row = new GenericRow();
+
+        while (segmentRecordReader.hasNext()) {
+          row = segmentRecordReader.next(row);
+          for (String multiValueColumn : multiValueColumns) {
+            multiValuesSum += ((Object[]) (row.getValue(multiValueColumn))).length;
+            numValues++;
+          }
+          row.clear();
+        }
+      } catch (Exception e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)